draft-ietf-masque-connect-ip.md

```diff
@@ -1271,6 +1271,58 @@ ADDRESS_ASSIGN capsule is parsed. This allows modifications to address
 assignement to operate atomically. Similarly, extensions that modify routing
 SHOULD behave similarly with regards to the ROUTE_ADVERTISEMENT capsule.
 
+# Performance Considerations
+
+Bursty traffic can often lead to temporally-correlated packet losses; in turn,
+this can lead to suboptimal responses from congestion controllers in protocols
+running inside the tunnel. To avoid this, endpoints SHOULD strive to avoid
+increasing burstiness of IP traffic; they SHOULD NOT queue packets in order to
+increase batching beyond the minimal amount required to take advantage of
+hardware offloads.
+
+When the protocol running inside the tunnel uses congestion control (e.g.,
+{{TCP}} or {{QUIC}}), the proxied traffic will incur at least two nested
+congestion controllers. The underlying HTTP connection MUST NOT disable
+congestion control unless it has an out-of-band way of knowing with absolute
+certainty that all of the inner packets belong to congestion-controlled
+connections.
+
+When the protocol running inside the tunnel uses loss recovery (e.g., {{TCP}}
+or {{QUIC}}), and the underlying HTTP connection runs over TCP, the proxied
+traffic will incur at least two nested loss recovery mechanisms. This can
+reduce performance as both can sometimes independently retransmit the same
+data. To avoid this, IP proxying SHOULD be performed over HTTP/3 to allow
+leveraging the QUIC DATAGRAM frame.
+
+## MTU Considerations
+
+When using HTTP/3 with the QUIC Datagram extension {{DGRAM}}, IP packets are
+transmitted in QUIC DATAGRAM frames. Since these frames cannot be fragmented,
+they can only carry packets up to a given length determined by the QUIC
+connection configuration and the Path MTU (PMTU). If an endpoint is using QUIC
+DATAGRAM frames and it attempts to route an IP packet through the tunnel that
+will not fit inside a QUIC DATAGRAM frame, the IP proxy SHOULD NOT send the IP
+packet in a DATAGRAM capsule, as that defeats the end-to-end unreliability
+characteristic that methods such as Datagram Packetization Layer PMTU Discovery
+(DPLPMTUD) depend on {{?DPLPMTUD=RFC8899}}. In this scenario, the endpoint
+SHOULD drop the IP packet and send an ICMP Packet Too Big message to the sender
+of the dropped packet; see {{Section 3.2 of ICMPv6}}.
+
+## ECN Considerations
+
+If a client or IP proxy with a connection containing an IP Proxying request
+stream disables congestion control, it MUST NOT signal Explicit Congestion
+Notification (ECN) {{!ECN=RFC3168}} support on that outer connection. That is,
```

Comment:
Due to the discussion of the next sentence I realised that there are issues with the first sentence in the  paragraph too. 

When a client establish a HTTP with a server it may not necessarily be known if the client will request Connect-IP service. That is the first issue. Thus, disabling ECN on the whole QUIC connection on any end-point that support Conenct-IP appears going to far. 

Secondly, this will be dependent on how the HTTP connection was established and how the HTTP datagrams are realized. So only over QUIC using QUIC datagrams is it possible to disable congestion control. And to disable ECN on just packets with QUIC datagrams leads to an additional complexity in tracking how the counter should change. This should be possible if the sender tracks the ECN correctly on a per packet basis and uses the ACK to determine which packet has been received at all and there are no erroneous re-marking of not-ECT to CE or either of the ECT. 
So I think potentially this is all up to the sender of how to deal with this. What is necessary to do is to (re-)mark each QUIC packet with an QUIC datagram containing a stream-id for which one have disabled congestion control to not-ECT. But potentially an implementor taking a short cut could disable sending ECT at the point it enables the congestion control bypass. However, that sender still needs to separate congestion tracking based on knowing if each sent packet should be counted or not when it receives the ACKs. 

But maybe this was intended with this first sentence. I would suggest that maybe the right change here is to write this paragraph as: 

If a client or IP proxy with a connection containing an IP Proxying request
 stream disables congestion control, it be required to not indicate  Explicit Congestion
 Notification (ECN) {{!ECN=RFC3168}} support on those packets in the outer connection.
That is, the QUIC sender MUST mark all IP headers with the Not-ECT codepoint for QUIC packets which are outside of congestion control. The endpoint SHALL continue to report ECN feedback via QUIC ACK_ECN frames or the TCP ECE bit if previously done in the QUIC connection, as the peer may not have disabled congestion control.

The change to the later sentence should fix the problem for the next comment too.

Reply:
Thanks, I've incorporated a lot of this into the PR.

The two things that I didn't incorporate were:
* DiffServ - I don't think we have good guidance to give so I'd rather avoid the topic
* The notion of "previously done on the connection" - I think the proposed phrasing handles your concern without going into overly specific guidance

Reply:
@gloinul are you OK with the text as-is now? It seems reasonable to me.

Reply:
I think the previously done part is fine to ignore. That is being solved. What I not happy yet about is the MUST NOT signal ECN in the first sentence. I think that is unclear what that normative statement refers to. So making that informal as I had in my proposed text is clearer, and only the normative about MUST mark not-ECT that is currently in the second sentence in the paragraph. Or are there any additional aspect that was targeted by that first MUST NOT?

Reply:
Replaced MUST NOT with cannot